go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,523 @@
+	for _, netName := range allNetworkNames {
+		nadController.perNetworkNadNameInfo.LockKey(netName)
+		nni, ok := nadController.perNetworkNadNameInfo.Load(netName)
+		if ok {
+			allNetworkControllers = append(allNetworkControllers, nni.oc)
+		}
+		nadController.perNetworkNadNameInfo.UnlockKey(netName)
+	}
+	return allNetworkControllers
+}
+
+// AddNetAttachDef adds the given nad to the associated controller. It creates the controller if this
+// is the first nad of the network.
+// Non-retriable errors (configuration error etc.) are just logged, and the function immediately returns nil.
+func (nadController *netAttachDefinitionController) AddNetAttachDef(ncm NetworkControllerManager,
+	netattachdef *nettypes.NetworkAttachmentDefinition, doStart bool) error {
+	var netConfInfo util.NetConfInfo
+	var nInfo util.NetInfo
+	var err, invalidNadErr error
+	var netName string
+
+	netAttachDefName := util.GetNadName(netattachdef.Namespace, netattachdef.Name)
+	klog.Infof("Add net-attach-def %s", netAttachDefName)
+
+	nInfo, netConfInfo, invalidNadErr = util.ParseNADInfo(netattachdef)
+	if invalidNadErr == nil {
+		netName = nInfo.GetNetworkName()
+		if netName == types.DefaultNetworkName {
+			invalidNadErr = fmt.Errorf("NAD for default network, skip it")
+		}
+	}
+
+	return nadController.perNadNetConfInfo.DoWithLock(netAttachDefName, func(nadName string) error {
+		nadNci, loaded := nadController.perNadNetConfInfo.LoadOrStore(nadName, &nadNetConfInfo{
+			NetConfInfo: netConfInfo,
+			netName:     netName,
+		})
+		if !loaded {
+			// first time to process this nad
+			if invalidNadErr != nil {
+				// invalid nad, nothing to do
+				klog.Warningf("net-attach-def %s is first seen and is invalid: %v", nadName, invalidNadErr)
+				nadController.perNadNetConfInfo.Delete(nadName)
+				return nil
+			}
+			klog.V(5).Infof("net-attach-def %s network %s first seen", nadName, netName)
+			err = nadController.addNadToController(ncm, nadName, nInfo, netConfInfo, doStart)
+			if err != nil {
+				klog.Errorf("Failed to add net-attach-def %s to network %s: %v", nadName, netName, err)
+				nadController.perNadNetConfInfo.Delete(nadName)
+				return err
+			}
+		} else {
+			klog.V(5).Infof("net-attach-def %s network %s already exists", nadName, netName)
+			nadUpdated := false
+			if invalidNadErr != nil {
+				nadUpdated = true
+			} else if nadNci.netName != netName {
+				// netconf network name changed
+				klog.V(5).Infof("net-attach-def %s network name %s has changed", netName, nadNci.netName)
+				nadUpdated = true
+			} else if !nadNci.CompareNetConf(netConfInfo) {
+				// netconf spec changed
+				klog.V(5).Infof("net-attach-def %s spec has changed", nadName)
+				nadUpdated = true
+			}
+
+			if !nadUpdated {
+				// nothing changed, may still need to start the controller
+				if !doStart {
+					return nil
+				}
+				err = nadController.addNadToController(ncm, nadName, nInfo, netConfInfo, doStart)
```

Comment:
in the case if NAD is updated (else case), we'd need to first delete the old NAD first before adding the new NAD to the network.

Reply:
if its possible I would like to do that. The thing is if someone updates a NAD lets say to have topology type from L3 to L2, we delete the L3 NAD, add the L2 one...now what happens to all the pods that were already attached? Looks like in the current code we would try to delete the logical switch, which would fail cause it has ports on it.

Reply:
as I understand the level driven queue implementation, also and I understand from a previous discussion with @maiqueb and @jcaamano, it is possible there is a race that we process the NAD in the worker thread, we might miss a deletion of old NAD if a subsequent new NAD with the same key comes right after. From the worker thread, this will be an update so we need to handle update of NAD anyway.

Reply:
Sure your explanation makes sense about treating it as level driven and all we know is a key needs to be processed, however we still need to reconcile what we had vs what we have now. So if we find the new NAD now has a different topology type, do we simply delete all the switches and router, and then recreate? Now any pods that were already created and attached to the old switch no longer have LSPs, but a new network exists? How do we solve this problem? Should we re-wire any existing pods?

Reply:
so perhaps the best solution is not to support NAD update (simple log an warning in onNetworkAttachDefinitionUpdate() without add the NAD to the worker queue)?